### PR TITLE
fix: strict mode for timestamp to block number conversion

### PR DESCRIPTION
**Pull request by akolotov**
## Motivation

During testing of PR #11163, it was discovered that the functionality for getting block numbers after specified timestamps (`Explorer.Chain.timestamp_to_block_number`) doesn't work as expected. Here are the issues found during investigation.

### 1. Unpredictable Results When Multiple Blocks Share the Same Timestamp

For chains that produce blocks with intervals under 1 second (like Arbitrum chains), multiple blocks can share the same timestamp. However, the `gt_timestamp_query` and `lt_timestamp_query` queries in the `Explorer.Chain.timestamp_to_block_number` function don't account for this scenario. Since each query returns only one block number, there's no guarantee which block will be selected from the set of blocks with `given_timestamp`, as the selection depends on the database's internal ordering behavior and query execution plan.

**Example:**

The following records exist in `Explorer.Chain.Block` , and `given_timestamp` set to "2025-01-08 22:04:45"

```
5005, 2025-01-08 22:04:47 <-- will be chosen by gt_timestamp_query
5004, 2025-01-08 22:04:44
5003, 2025-01-08 22:04:44
5002, 2025-01-08 22:04:44 <-- will be chosen by lt_timestamp_query
```

### 2. Request Condition Violation

`Explorer.Chain.timestamp_to_block_number` executes `lt_timestamp_query |> subquery() |> union(^gt_timestamp_query)` to find the block closest to `given_timestamp`. This can lead to two incorrect scenarios:

- A block with a higher number is chosen when `:before` is passed as the parameter `closest`
- A block with a lower number is chosen when `:after` is passed as the parameter `closest`

Next, the chosen block is passed to `Explorer.Chain.get_block_number_based_on_closest`, which invokes either `Explorer.Chain.BlockNumberHelper.previous_block_number` or `Explorer.Chain.BlockNumberHelper.next_block_number`, depending on the case.

```
..Explorer.Chain.timestamp_to_block_number
....Explorer.Chain.get_block_number_based_on_closest
......Explorer.Chain.BlockNumberHelper.{previous_block_number,next_block_number}
........Explorer.Chain.BlockNumberHelper.neighbor_block_number
..........Explorer.Chain.BlockNumberHelper.move_by_one
```

These functions simply increment or decrement the block number based on the direction (`previous` or `next`), without verifying the timestamps of the resulting blocks.

Since `gt_timestamp_query` and `lt_timestamp_query` don't guarantee returning subsequent blocks (as noted in the first issue), `Explorer.Chain.get_block_number_based_on_closest` might return a block that violates the condition specified in the `closest` parameter.

For instance, in the example above where `closest` is `after`, block `5003` will be returned even though its timestamp is **before** the given timestamp.

### 3. Non-indexed Block

Consider another example where the database has only two records (blocks `5006` and `5007` are not yet indexed), `given_timestamp` is "2025-01-08 22:04:50", and `closest` is `after`:

```
5008, 2025-01-08 22:04:55 <-- will be chosen by gt_timestamp_query
5005, 2025-01-08 22:04:47 <-- will be chosen by lt_timestamp_query
```

The function `Explorer.Chain.timestamp_to_block_number` will return `5006` even though this block isn't in the Blockscout database. This occurs because after selecting candidate blocks from `Explorer.Chain.Block` using `gt_timestamp_query` and `lt_timestamp_query`, the adjustment process described in issue #2 doesn't verify whether the calculated block number exists in `Explorer.Chain.Block`.

## Proposed Solution

The first issue is resolved by including the `number` field of `Explorer.Chain.Block` in the ordering instructions for queries that identify blocks closest to the given timestamp.

The behaviors described in the second and third issues may be intentional, so it was decided to separate the current behavior from the behavior needed for changes in PR #11163. A new parameter `strict` to the `timestamp_to_block_number` function is introduced. When set to `true`, the function selects either `gt_timestamp_query` or `lt_timestamp_query` based on the direction specified in `closest`. This ensures the function will either return an indexed block that matches the `closest` requirement or return `:not_found` if no blocks exist before or after the given timestamp.

## Changelog

1. Since the `Explorer.Chain` module exceeds 5,500 lines of code and smaller modules are preferable (see explanation here: [`apps/explorer/lib/explorer/chain/arbitrum/reader/README.md`](https://github.com/blockscout/blockscout/blob/7e522eb194dba4e141ad9b5ea965cb1e3c47f66b/apps/explorer/lib/explorer/chain/arbitrum/reader/README.md)), it is decided to move the functions `timestamp_to_block_number` and `get_block_number_based_on_closest` to a new module `Explorer.Chain.Block.Reader.General` (`apps/explorer/lib/explorer/chain/block/reader/general.ex`).

2. Since the `gt_timestamp_query` and `lt_timestamp_query` queries share much functionality, they are implemented through three functions: `build_directional_query`, `adjust_query_to_highest_block`, and `adjust_query_to_lowest_block`. The `direction` parameter passed to `build_directional_query` determines which query to compose. `build_directional_query` creates the common query structure, while `adjust_query_to_highest_block` and `adjust_query_to_lowest_block` add conditions for selecting blocks before or after the given timestamp and define the element ordering. As described in the "Proposed Solution" section, the ordering instructions now include block numbers.

3. The boolean parameter `strict` is added to the `Explorer.Chain.Block.Reader.General.timestamp_to_block_number function` specification. The previous implementation of `timestamp_to_block_number` remains under the clause `timestamp_to_block_number(given_timestamp, closest, from_api, false)`. The clause `timestamp_to_block_number(given_timestamp, closest, from_api, true)` composes one query through `build_directional_query` based on `closest`, runs the query using the repo according to `from_api`, and handles the response as follows:
    - if it's `nil`, return `{:error, :not_found}`
    - if a block number is found, return it.

4. It is assumed that if the `strict` parameter is not passed to the function, its default value will be `false`. Therefore, besides interacting with the new module, only minimal changes are needed in these modules:
    - `Explorer.Chain.AdvancedFilter` (`apps/explorer/lib/explorer/chain/advanced_filter.ex`)
    - `BlockScoutWeb.API.RPC.BlockController` (`apps/block_scout_web/lib/block_scout_web/controllers/api/rpc/block_controller.ex`)
    - `Explorer.Chain.Transaction.History.Historian` (`apps/explorer/lib/explorer/chain/transaction/history/historian.ex`)

These changes only involve updating the module reference for calling `timestamp_to_block_number`.

Additionally, the function `Indexer.Fetcher.Arbitrum.Utils.Db.Common.closest_block_after_timestamp` in `apps/indexer/lib/indexer/fetcher/arbitrum/utils/db/common.ex` has been modified to call `timestamp_to_block_number` with the `strict` parameter set to `true`.

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [ ] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/setup/env-variables/README.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/setup/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/setup/env-variables).
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Release Notes

- **New Features**
  - Introduced a new module for enhanced timestamp-to-block number conversion.
  - Added functionality for managing null rounds in Filecoin chains.
  - New test module added to validate timestamp-to-block number functionality.

- **Improvements**
  - Enhanced block number retrieval methods for greater flexibility.
  - Improved documentation for better understanding and usability.

- **Technical Updates**
  - Updated module dependencies for block number and timestamp lookups.
  - Added new utility functions for managing block numbers and null rounds.
  
- **Bug Fixes**
  - Removed outdated test validations for deprecated functions.

### CODE REVIEW

File: apps/explorer/lib/explorer/chain/block/reader/general.ex [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    BlockNumberHelper
+  }
+
+  alias Explorer.Repo
+
+  require Logger
+
+  @doc """
+    Finds the block number closest to a given timestamp, optionally adjusting
+    based on whether the block should be before or after the timestamp.
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex]: Thanks! Addressed in d844dfd

**Comment by akolotov** [comment]
@vbaranov @sl1depengwyn @fedor-ivn please take a look

File: apps/explorer/test/explorer/chain_test.exs [hunk 2]
```diff
@@ -4353,9 +4354,9 @@ defmodule Explorer.ChainTest do
       block = insert(:block, timestamp: timestamp)
       expected = {:ok, block.number}
 
-      assert ^expected = Chain.timestamp_to_block_number(timestamp, :after, true)
```
Comment (vbaranov) [inline apps/explorer/test/explorer/chain_test.exs:4356]: `timestamp_to_block_number/3` test suite shouldn't exist in `apps/explorer/test/explorer/chain_test.exs` since the function migrated to `BlockGeneralReader` module. Please move this test to a new file (`apps/explorer/test/explorer/chain/block/reader/general_test.exs`) with tests specific to `BlockGeneralReader` module.

File: apps/explorer/lib/explorer/chain/block/reader/general.ex [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    - `{:ok, block_number}` where `block_number` is the block number closest to
+      the specified timestamp.
+    - `{:error, :not_found}` if no block is found within the specified criteria
+      or if the found block represents a null round.
+  """
+  @spec timestamp_to_block_number(DateTime.t(), :before | :after, boolean(), boolean()) ::
+          {:ok, Block.block_number()} | {:error, :not_found}
+  def timestamp_to_block_number(given_timestamp, closest, from_api, strict \\ false)
+
+  def timestamp_to_block_number(given_timestamp, closest, from_api, true) do
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/block/reader/general.ex:53]: Let's add a unit test of `timestamp_to_block_number/4` function with `string=true` option.

File: apps/explorer/test/explorer/chain_test.exs (hunk 2 above)
Comment (vbaranov) [reply apps/explorer/test/explorer/chain_test.exs:4356]: Also, `timestamp_to_block_number` has 4 arguments now. Let's rename test suite as well.

File: apps/explorer/lib/explorer/chain/block/reader/general.ex [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  #
+  # ## Returns
+  # - A query that will find the closest consensus block in the specified direction
+  @spec build_directional_query(DateTime.t(), :before | :after) :: Ecto.Query.t()
+  defp build_directional_query(timestamp, direction) do
+    base_query =
+      from(block in Block,
+        where: block.consensus == true,
+        limit: 1,
+        select: block
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/block/reader/general.ex:109]: This could be omitted. By default, all fields are returned.

File: apps/explorer/lib/explorer/chain/block_number_helper.ex [hunk 5]
```diff
@@ -1,26 +1,115 @@
...
+      # only blocks with consensus `true` must be taken into account here as well.
       defp neighbor_block_number(number, direction), do: move_by_one(number, direction)
+
+      # For non-Filecoin chains, the concept of null rounds doesn't exist, so it
+      # is assumed that the block number is always valid.
+      defp do_find_next_non_null_round_block(block_number, _), do: {:ok, block_number}
   end
 
+  @spec move_by_one(non_neg_integer(), :previous | :next) :: non_neg_integer()
   def move_by_one(number, :previous), do: number - 1
```
Comment (fedor-ivn) [inline apps/explorer/lib/explorer/chain/block_number_helper.ex:113]: Shouldn't we somehow handle the case when given `number` is `0`?

Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/block_number_helper.ex:113]: It might not make much sense, but I wanted to point it out just in case.

File: apps/explorer/lib/explorer/chain/block/reader/general.ex [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    - `given_timestamp`: The timestamp for which the closest block number is
+      being sought.
+    - `closest`: A direction indicator (`:before` or `:after`) specifying
+                whether the block number returned should be before or after the
+                given timestamp.
+    - `from_api`: A boolean flag indicating whether to use the replica database
+                  or the primary one for the query.
+    - `strict`: A boolean flag controlling the block selection behavior:
+      * `false` (default): Returns the block with smallest absolute time
+        difference, then adjusts based on `closest` parameter if needed
```
Comment (fedor-ivn) [inline apps/explorer/lib/explorer/chain/block/reader/general.ex:39]: Am I correct in understanding that (in case `strict` = `false`) if the block number adjusted by the `closest` parameter doesn’t exist in the database, `{:ok, block_number}` would still be returned?

File: apps/explorer/test/explorer/chain_test.exs
Comment (fedor-ivn) [inline apps/explorer/test/explorer/chain_test.exs:1]: Would it make sense to add tests specifically for the Filecoin case?

File: apps/explorer/lib/explorer/chain/null_round_height.ex [hunk 7]
```diff
@@ -67,15 +133,116 @@ defmodule Explorer.Chain.NullRoundHeight do
...
+    |> case do
+      nil ->
+        # If no valid block found in current batch, try the next batch
+        last_block_number = List.last(existing_blocks)
+        process_next_batch(last_block_number, direction)
+
+      number ->
+        {:ok, number}
+    end
   end
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/null_round_height.ex:247]: As far as I understand if there is a null round block then there is no consensus block with such number, so I suppose this function does not need both `existing_blocks` and `null_rounds`, i.e. condition `not MapSet.member?(null_rounds_set, number)` is always true, this function can be implemented simmilar to `find_neighbor_from_previous`

File: apps/explorer/lib/explorer/chain/block/reader/general.ex [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+  def timestamp_to_block_number(given_timestamp, closest, from_api, true) do
+    query = build_directional_query(given_timestamp, closest)
+
+    case select_repo(api?: from_api).one(query, timeout: :infinity) do
+      nil ->
+        {:error, :not_found}
+
+      %{number: number} ->
+        direction = if closest == :before, do: :previous, else: :next
+        BlockNumberHelper.find_next_non_null_round_block(number, direction)
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/block/reader/general.ex:62]: Consensus block number returned from blocks table can't be null round

**Comment by sl1depengwyn** [comment]
Also, just to understand if I get the PR right, as far as I see changes in this PR solves both first issues, but this is contradicts with PR description

> The first issue is resolved by including the number field of Explorer.Chain.Block in the ordering instructions for queries that identify blocks closest to the given timestamp.

Am I right?

**Comment by sl1depengwyn** [comment]
> Also, just to understand if I get the PR right, as far as I see changes in this PR solves both first issues, but this is contradicts with PR description
> 
> > The first issue is resolved by including the number field of Explorer.Chain.Block in the ordering instructions for queries that identify blocks closest to the given timestamp.
> 
> Am I right?

Ah, if we take into account non-indexed blocks, then the second issue can still appear in non strict mode

File: apps/explorer/lib/explorer/chain/block/reader/general.ex
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:1]: Thank you for raising this question about the module organization. Let me explain the reasoning behind the current structure:

1. The `reader/` directory was introduced as a planned home for block-related DB reading functionality that needs to be moved from `Explorer.Chain` and other places. This follows our goal of maintaining module sizes between 500-800 LOC maximum, as outlined in the [Arbitrum reader README](https://github.com/blockscout/blockscout/blob/7e522eb194dba4e141ad9b5ea965cb1e3c47f66b/apps/explorer/lib/explorer/chain/arbitrum/reader/README.md).

2. While moving everything to a single `apps/explorer/lib/explorer/chain/block/reader.ex` might seem simpler now, it would likely exceed our desired module size in the near future as we continue moving block-related DB operations from `Explorer.Chain` (which currently exceeds 5,500 LOC).

3. The current structure with a `reader/` directory:
   - Provides a clear location for block-related DB operations
   - Allows gradual, organized migration of functions from `Explorer.Chain`
   - Maintains module size constraints proactively
   - Supports collaborative development with smaller, focused files
   - Optimizes for LLM-based development by maintaining clear context windows
   - Reduces potential merge conflicts

4. The term "General" in the current module name could be more specific (e.g., `timestamp.ex`) to better reflect its functionality. However, having it under the `reader/` directory sets a good precedent for organizing future block-reading modules as we continue to break down `Explorer.Chain`.

This approach tries to be one step ahead, avoiding the need for additional splitting operations in the future while maintaining the benefits outlined in the Arbitrum reader documentation regarding collaborative and LLM-based development.

File: apps/explorer/lib/explorer/chain/block_number_helper.ex (hunk 5 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block_number_helper.ex:113]: After analyzing the codebase, I found that `move_by_one(0, :previous)` is not possible in practice because:

1. The function is only called through `neighbor_block_number/2` which is used by `previous_block_number/1` and `next_block_number/1`
2. The main usage is in `get_block_number_based_on_closest/4` which only calls `previous_block_number` when a block's timestamp is greater than the target timestamp
3. Since block 0 (genesis) would have the earliest timestamp, it would never need to go to a previous block

While we could add a guard for block 0, it would be defensive programming for a case that can't occur in the current codebase architecture.

File: apps/explorer/lib/explorer/chain/block/reader/general.ex (hunk 6 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:39]: Yes, you are correct. When `strict=false`, the function will return `{:ok, block_number}` even if the adjusted block number doesn't exist in the database. 

It's important to note that this is not a new behavior - this is exactly how `timestamp_to_block_number` functioned before the changes in this PR. The `strict=false` option preserves this original behavior to maintain backward compatibility with any code that might rely on it.

The behavior occurs because:
1. The function first finds the block with the closest timestamp in the database
2. Then it adjusts that block number based on the `closest` parameter using `BlockNumberHelper.previous_block_number/1` or `BlockNumberHelper.next_block_number/1`
3. The adjustment is purely arithmetic (±1) and doesn't verify the existence of the resulting block number
4. Returns `{:ok, block_number}` with the adjusted number, regardless of whether it exists

The new `strict=true` mode was added as an alternative when you need to ensure the returned block number exists in the database, but the original behavior remains unchanged when `strict=false` (which is the default).

File: apps/explorer/lib/explorer/chain/null_round_height.ex (hunk 7 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/null_round_height.ex:247]: Thank you for your comment about the potential redundancy in the `find_first_valid_block` function. You mentioned "As far as I understand if there is a null round block then there is no consensus block with such number". Could you help me understand if it is handled during the indexing process? May you direct me to the specific lines of the code?

The current module documentation describes null rounds in Filecoin, but I want to ensure I understand how Blockscout specifically handles them during indexing before making any assumptions about mutual exclusivity with consensus blocks.

File: apps/explorer/lib/explorer/chain/block/reader/general.ex (hunk 8 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:62]: Could you help me understand if it is handled during the indexing process? May you direct me to the specific lines of the code where it is guaranteed?

**Comment by akolotov** [comment]
> > Also, just to understand if I get the PR right, as far as I see changes in this PR solves both first issues, but this is contradicts with PR description
> > > The first issue is resolved by including the number field of Explorer.Chain.Block in the ordering instructions for queries that identify blocks closest to the given timestamp.
> > 
> > 
> > Am I right?
> 
> Ah, if we take into account non-indexed blocks, then the second issue can still appear in non strict mode

Yes, you're correct in your follow-up observation. While the PR changes do affect both issues, in non-strict mode (`strict=false`) the second issue can still occur with non-indexed blocks. This is intentional to maintain backward compatibility with the existing behavior. The strict mode (`strict=true`) was introduced specifically to provide a way to avoid these issues when needed.

File: apps/explorer/test/explorer/chain_test.exs
Comment (akolotov) [reply apps/explorer/test/explorer/chain_test.exs:1]: To write comprehensive tests for the Filecoin case, I would need some guidance on:
1. How to properly set up the test environment for Filecoin chain type
2. What are typical patterns of null rounds we should test for
3. Any specific edge cases in Filecoin that should be covered

Could you provide some examples or point me to existing Filecoin-specific tests that could serve as a reference?

File: apps/explorer/lib/explorer/chain/null_round_height.ex [hunk 9]
```diff
@@ -67,15 +133,116 @@ defmodule Explorer.Chain.NullRoundHeight do
     end
   end
 
+  @spec move_by_one(non_neg_integer(), :previous | :next) :: non_neg_integer()
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/null_round_height.ex]: addressed in 17f7d2b

File: apps/explorer/lib/explorer/chain/block/reader/general.ex [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+          Block.block_number()
+  defp get_block_number_based_on_closest(closest, timestamp, given_timestamp, number) do
+    # Note: When calculating adjacent block numbers (previous/next) for both `:before`
+    # and `:after` cases, there is no guarantee that the calculated block number
+    # exists in the database
+
+    case closest do
+      :before ->
+        if DateTime.compare(timestamp, given_timestamp) == :lt ||
+             DateTime.compare(timestamp, given_timestamp) == :eq do
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex]: Addressed in c3fd6a7

File: apps/explorer/lib/explorer/chain/block/reader/general.ex (hunk 4 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex]: Addressed in c3fd6a7

File: apps/explorer/lib/explorer/chain/block/reader/general.ex (hunk 3 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:53]: Addressed in f0b87b7

File: apps/explorer/test/explorer/chain_test.exs (hunk 2 above)
Comment (akolotov) [reply apps/explorer/test/explorer/chain_test.exs:4230]: Addressed in f0b87b7

File: apps/explorer/lib/explorer/chain/block/reader/general.ex
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:1]: I agree that splitting the massive `Explorer.Chain` module into more specialized parts makes sense. However, as I get it, our usual naming convention is to keep “general” functionality in a top-level module (e.g., `Explorer.Chain.Block.Reader`) and then place more specialized modules under that (e.g., `Explorer.Chain.Block.Reader.Arbitrum` or `Explorer.Chain.Block.Reader.Timestamp`).

For instance, we have `Explorer.Chain.Block` at `apps/explorer/lib/explorer/chain/block.ex`, and then more specific block-related logic in `Explorer.Chain.Block.Reward` at `apps/explorer/lib/explorer/chain/block/reward.ex`. Some might argue that this structure complicates discovering modules via a file explorer in IDE, but in practice, I personally use advanced search or ElixirLS capabilities, and each logical piece remains in a consistent hierarchy. Do you see any particular downside to this approach?

File: apps/explorer/lib/explorer/chain/block_number_helper.ex (hunk 5 above)
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/block_number_helper.ex:113]: While it may not occur in the current codebase, the function isn’t private—meaning it could be used in other, unpredictable contexts. Without deeper insight into the `BlockNumberHelper` module, a developer might not realize there’s a caveat with the genesis block.

File: apps/explorer/lib/explorer/chain/block/reader/general.ex
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:1]: If we follow the approach of having a single main module with multiple specialized submodules, there will always be a strong temptation for developers to add new functionality to the main module rather than carefully placing it in the appropriate submodule. A real-world example of this issue is `Explorer.Chain`. Despite having multiple child modules under `Explorer.Chain`, this module somehow still grew to 5,500 LOC - plus an additional 4,200 LOC in `/workspace/apps/explorer/test/explorer/chain_test.exs`.

[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:1]: > What I see is that this is a matter of preferences rather than standards.

Exactly. In this case, it reflects the preferences of the contributors who worked on Blockscout before, rather than a strict standard I’m enforcing.

[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:1]: Are you sure that having a module at `XXX.Reader` instead of `XXX.Reader.General` actually solves the issue you mentioned? What would prevent someone from putting everything into `XXX.Reader.General` anyway? I’m trying to understand how the location alone addresses the concern.

File: apps/explorer/lib/explorer/chain/null_round_height.ex (hunk 7 above)
Comment (sl1depengwyn) [reply apps/explorer/lib/explorer/chain/null_round_height.ex:243]: Filecoin node returns error when null round block is requested, example:
```bash
curl --location 'https://filecoin.lava.build' \
--header 'Content-Type: application/json' \
--data '{
    "jsonrpc": "2.0",
    "method": "eth_getBlockByNumber",
    "params": [
        "0x453361",
        true
    ],
    "id": 1
}'
{"jsonrpc":"2.0","id":1,"error":{"code":1,"message":"requested epoch was a null round"}}%                   
```
So, we have no block hash, for instance, to import something to blocks table, on such errors we are simply populate null_round_heights table and remove such blocks from missed ranges here https://github.com/blockscout/blockscout/blob/master/apps/indexer/lib/indexer/block/catchup/fetcher.ex#L233

File: apps/explorer/lib/explorer/chain/block/reader/general.ex
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:1]: The idea is that if we immediately create `XXX.Reader.YYY` without having a `XXX.Reader` module, developers won’t be tempted to leave code in `XXX.Reader`. Reviewers of future changes are also unlikely to enforce size constraints on `XXX.Reader`, which could lead to the same issue over time. That’s why I suggest structuring it as submodules from the start, avoiding the "first a single module, then we’ll split it if necessary" phase.  
  
If the concern is mainly about naming, the discussed module could be renamed to `Reader.Timestamp`. However, I’m not in favor of this approach because, at this point, it's unclear whether additional functionality will be introduced in this module.

File: apps/explorer/lib/explorer/chain/null_round_height.ex (hunk 7 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/null_round_height.ex]: Thanks. It clear now. In 28ea1c8 I removed handling of null rounds for the strict mode.

File: apps/explorer/lib/explorer/chain/block/reader/general.ex (hunk 8 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex]: The same as https://github.com/blockscout/blockscout/pull/11633#discussion_r1943558616

File: apps/explorer/test/explorer/chain_test.exs
Comment (akolotov) [reply apps/explorer/test/explorer/chain_test.exs:1]: Do you think it is still applicable after 28ea1c8?

File: apps/explorer/lib/explorer/chain/block_number_helper.ex (hunk 5 above)
Comment (akolotov) [reply apps/explorer/lib/explorer/chain/block_number_helper.ex]: Agree. But I extended the functionality to not go below `FIRST_BLOCK` configured for the application.Take a look at 9e1f3c9.  wdyt?

**Review by sl1depengwyn — APPROVED** [review]

**Review by vbaranov — APPROVED** [review]

File: apps/explorer/lib/explorer/chain/block/reader/general.ex
[the remark this answers is not in the document]
Comment (fedor-ivn) [reply apps/explorer/lib/explorer/chain/block/reader/general.ex:1]: Well, I think I understand your point now. You mean that once we add more functionality, we can rename the module and avoid having a “general” module altogether, right? If so, that makes sense to me.

**Review by fedor-ivn — APPROVED** [review]

File: apps/explorer/test/explorer/chain_test.exs
Comment (fedor-ivn) [reply apps/explorer/test/explorer/chain_test.exs:1]: No, I think it's fine now
